File: Week_03/at-a-conference/src/Conference.js
class Conference {
  constructor(name, maxRegistrants, attendees = [], sessions = []) {
    this.name = name;
    this.maxRegistrants = maxRegistrants;
    this.attendees = attendees;
    this.sessions = sessions;
  };

  register(person) {
    if (this.attendees.length < this.maxRegistrants && 
      !this.attendees.some(attendee => attendee.email === person.email)) {
      this.attendees.push(person);
      return true;
    };
    return false;
  };

  addSession(session) {
    if (this.attendees.some(attendee => attendee.email === session.facilitator.email)){
      this.sessions.push(session);
      return true;
    }
    return false;
  };

  summary() {
    let pamphlet = `Conference name: ${this.name}\n` + 
                  `Number of registrants: ${this.attendees.length}`;
    this.attendees.forEach(attendee => {
      pamphlet += `\n  ${attendee.firstName} ${attendee.lastName}`;
    });
    pamphlet += `\nNumber of sessions: ${this.sessions.length}`;
    this.sessions.forEach(session => {
      pamphlet += `\n  Title: ${session.title}\n  Facilitator: ${session.facilitator}`;
    });
    pamphlet += `\nRegistration is ${(this.attendees.length < this.maxRegistrants) ? `open (${this.maxRegistrants - this.attendees} seats remaining)`: `closed`}`;
    // if (this.attendees.length < this.maxRegistrants) {
    //   //
    // };
    return pamphlet;
  };
};

export default Conference;

File: Week_11/java-at-a-conference/src/main/java/Conference.java
import java.util.ArrayList;

public class Conference {
  private String name;
  private int maxRegistrants;
  private ArrayList<Person> attendees;
  private ArrayList<Session> sessions;

  public Conference(String name, int maxRegistrants) {
    this.name = name;
    this.maxRegistrants = maxRegistrants;
    this.attendees = new ArrayList<>();
    this.sessions = new ArrayList<>();
  }

  public String getName() {
    return this.name;
  }

  public int getMaxRegistrants() {
    return this.maxRegistrants;
  }

  public ArrayList getAttendees(){
    return this.attendees;
  }

  public ArrayList getSessions(){
    return this.sessions;
  }

  public boolean isOpen(){ return this.attendees.size() < this.maxRegistrants; }

  public boolean register(Person newAttendee){
    ArrayList<String> attendeeEmails = new ArrayList<>();
    for (Person attendee : this.attendees) {
      attendeeEmails.add(attendee.getEmail());
    }
    if (attendeeEmails.contains(newAttendee.getEmail()) || !isOpen()) {  return false; }
    else {
      this.attendees.add(newAttendee);
      return true;
    }
  }

  public boolean addSession(Session session) {
    this.sessions.add(session);
    return true;
  }

  public String getSummary() {
    String output = "Welcome to the \"" + getName() + "\" Conference";
    output += "\nNumber of Attendees: " + this.attendees.size() + "/" + getMaxRegistrants();
    for(Person person : this.attendees) {
      output += "\n  " + person.getName();
    }
    output += "\nNumber of Sessions: " + this.sessions.size();
    for(Session session : this.sessions) {
      output += "\n  \"" + session.getName() + "\", facilitated by: " + session.getFacilitator().getName();
    }
    if (isOpen()) { output += "\nRegistration is still open.\n"; }
    else { output += "\nRegistration is now closed.\n"; }
    return output;
  }
}
